docs/C++.md

```diff
@@ -15,3 +15,17 @@ for multiple dispatch in the IR and Visitor classes.
 C++ does not handle inner classes natively, so to implement an inner class,
 we use a nested class with a reference member to the containing class
 called `self`.
+
+### RTTI
+IR nodes and their handling code (e.g. Visitors) make extensive use of RTTI to perform type-checking and corresponding downcasting from `Node*` down to a particular implementation. C++ native RTTI as implemented in `dynamic_cast` and `typeid` built-ins is inherently slow and has lots of overhead. To circumvent this IR node classes use dedicated light-weight RTTI designed for semi-open class hierarchies. The implementation itself requires some boilerplate that is automatically generated by `ir-generator` for all IR classes from `.def`. The corresponding RTTI functionality could be accessed via `ICastable` interface from `lib/castable.h`. The lower-level RTTI implementation details are in `lib/rtti.h`. Overall, use of `dynamic_cast` and `typeid` for IR nodes is discouraged.
+
+#### Using lightweight RTTI for other class hierarchies
+Other class hierarchies besides `IR::Node` descendants could also benefit from lightweight RTTI functionality. In order to use it one needs to annotate the intended class hierarchy as follows:
+  * Derive from `IR::Castable` (or lower-level `RTTI::Base`)
+  * Use `DECLARE_TYPEINFO(Class, Bases..)` variadic macro to define necessary boilerplate. Here typeid will generated during the compile time. Should necessary, explicit typeid could be specified by `DECLARE_TYPEINFO(Class, TypeId, Bases...)`
```

Comment:
Generally not a fan of forced macro usage. It makes the compiler much harder to use or extend. Any way we can avoid this using C++17/20? I will try to take a closer look later.

Reply:
The current code relies on C++17. Macro is used to intrusively inject the code to the class hierarchy. An alternative would be a CRTP-like mixin. However, IMO this would both negatively affect the compile time. Please see the particular implementation at https://github.com/p4lang/p4c/pull/4377/files#diff-d0caf68249685b4c35c4e2aa7e95735b438987527ecdaf4db8f29a8158c8a405R167